### PR TITLE
[new release] mirage, mirage-runtime, functoria and functoria-runtime (4.0.0+beta1)

### PR DESCRIPTION
**Pull request by dinosaure**
The MirageOS library operating system

- Project page: <a href="https://github.com/mirage/mirage">https://github.com/mirage/mirage</a>
- Documentation: <a href="https://mirage.github.io/mirage/">https://mirage.github.io/mirage/</a>

##### CHANGES:

Refactor build process to use [Dune](https://dune.build/) build system. The
motivation is to drop `ocamlbuild`-induced technical debt and to obtain
first-class support for _cross-compilation_. To learn more about how Dune is
able to perform cross-compilation, please refer to the
[documentation](https://dune.readthedocs.io/en/stable/cross-compilation.html).

Main changes:

* Two opam files are generated when running `mirage configure`:
  - `<unikernel>-switch.opam`: for dependencies that are meant to be installed
    in the user's opam switch. It comprises of build tools such as
    `ocaml-freestanding` for Solo5 targets.
  - `<unikernel>-monorepo.opam`: for unikernel dependencies, they are locally
    fetched to compose a _dune workspace_.

* Unikernel dependencies are fetched in the source project using the
  `opam-monorepo` tool. This tool reads the `<unikernel>-monorepo.opam` file and
  make use of the opam solver to compute the transitive dependency set, saves
  that as a _lockfile_, and fetch sources in the `duniverse/` subfolder.
  More info on the [Github repository](https://github.com/ocamllabs/opam-monorepo).

* The compilation scheme use `dune`'s concept of a _workspace_: it's a set of
  libraries that are built together using the same _context_. For each
  compilation target, the Mirage tool is able to generate a _context_ definition
  able to compile for that target. A _context_ is defined by an OCaml compiler
  (or cross-compiler) toolchain, as defined by `findlib`, it can be tuned with
  environment variables and custom flags for the OCaml or underlying C compiler.

* The usual workflow: `mirage configure && make depends && mirage build` does
  not change however. Generated files are located into the new directory
  `./mirage` (OPAM files, `main.ml`, `key_gen.ml` or `manifest.json`) and
  the final artifact is created into the `./dist` directory.

Breaking changes:

* Unikernel dependencies need to use `dune` as a build system. Other build
  systems may be sandboxed in a Dune package, but the recommended way is to
  switch to `dune`. A set of packages have been ported to dune but the changes
  are not upstreamed, they are available in the `dune-universe`
  [overlay repository](https://github.com/dune-universe/opam-overlays). A
  specific package must be _dunified_ in a _certain_ way to allow the
  cross-compilation: `zarith`. It is maitained by the MirageOS core team and
  available into our own
  [mirage overlay repository](https://github.com/dune-universe/mirage-opam-overlays).
  These repositories are configured by default.
  To remove them, use `mirage configure --no-extra-repo`.

* `Functoria_runtime.info` and `Mirage_runtime.info` now list all the libraries
  that are statically linked against the unikernel. The `packages` field have
  been removed and the `libraries` field is now accurate and contains the
  versions computed by `dune-build-info`.

* The DSL to describe devices into the `config.ml` was updated. We don't use
  anymore objects and we replace it by an usage of `Mirage.impl` which expects
  same _fields_ than before.

### CODE REVIEW

**Comment by kit-ty-kate** [comment]
The version number doesn’t seem correct for a beta release as: `4.0.0 < 4.0.0+beta1`
What you want is `4.0.0~beta1`.

See https://opam.ocaml.org/doc/Manual.html#version-ordering

**Comment by kit-ty-kate** [comment]
Also, given these are beta versions. Could you add:
```
flags: avoid-version
```
This is opam 2.1 only. So maybe it would be good to make these available only for opam 2.1 or do the same thing as for ocaml-variants and require the user to use https://github.com/ocaml/ocaml-beta-repository when `opam-version < "2.1.0"`. See https://github.com/ocaml/opam-repository/pull/20519 for example

**Comment by dinosaure** [comment]
Currently, we should constraint the user to us only `opam-version >= "2.1.0"`. Where I should add it into my OPAM files?

**Comment by kit-ty-kate** [comment]
>  EDIT: about the hidden-flags, should I put it upstream?

no, it does not make sense outside of opam-repository.

> Currently, we should constraint the user to us only `opam-version >= "2.1.0"`. Where I should add it into my OPAM files?

As pointed above you can either reuse the `ocaml-beta` package (see ocaml-variants definition for examples) or do:
```
available: opam-version >= "2.1.0"
```
but as you wish.

**Comment by dinosaure** [comment]
Currently, I think this PR is fine but the lower-bounds fails on MirageOS 3 packages (which is, _a priori_, unrelated to this PR). I can make an issue to keep in our mind that MirageOS 3 has some issues about lower-bounds but can we consider this PR as ready to merge?

**Comment by kit-ty-kate** [comment]
For the record, could you give the error messages that prompted https://github.com/ocaml/opam-repository/pull/20558/commits/4fc8d2be3c0b2a6c18c503f6e4634a93201c4d17 ?

**Comment by dinosaure** [comment]
> For the record, could you give the error messages that prompted 4fc8d2b ?

I pushed-force to add an explanation on this specific commit.

**Comment by kit-ty-kate** [comment]
It would be nice to also have the full error message we can make sure that this new dependency isn’t there just because it’s needed in the tests or some other possible errors. See for example: https://github.com/ocaml/opam-repository/pull/20544

**Comment by hannesm** [comment]
The failures may be fixed by using `< "4.0"` instead of `< "4.0.0"` in the constraints of earlier mirage releases. The reasoning:
```OCaml
# OpamVersionCompare.compare  "4.0.0" "4.0.0~beta1";;
- : int = 1
# OpamVersionCompare.compare  "4.0" "4.0.0~beta1";;
- : int = -1
```

File: packages/conan-cli/conan-cli.0.0.1/opam [hunk 1]
```diff
@@ -20,7 +20,7 @@ depends: [
   "crowbar"        {with-test}
   "fmt"            {with-test}
   "rresult"        {with-test}
-  "mirage"              {with-test}
+  "mirage"              {with-test & < "4.0.0"}
```
Comment (hannesm) [inline packages/conan-cli/conan-cli.0.0.1/opam:23]:
```suggestion
  "mirage"              {with-test & < "4.0"}
```

File: packages/mirage/mirage.3.1.1/opam [hunk 2]
```diff
@@ -18,7 +18,7 @@ depends: [
   "ocaml" {>= "4.04.2"}
   "jbuilder" {>= "1.0+beta10"}
   "ipaddr" {>= "2.6.0" & < "5.0.0"}
-  "functoria" {>= "2.2.0"}
+  "functoria" {>= "2.2.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.1.1/opam:21]:
```suggestion
  "functoria" {>= "2.2.0" & < "4.0"}
```

File: packages/mirage/mirage.3.10.1/opam [hunk 3]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.10.1/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.10.2/opam [hunk 4]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "2.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.10.2/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.10.3/opam [hunk 5]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "2.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.10.3/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.10.4/opam [hunk 6]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "2.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.10.4/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.10.6/opam [hunk 7]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "2.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.10.6/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.10.7/opam [hunk 8]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "2.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.10.7/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.2.0/opam [hunk 9]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.04.2"}
   "jbuilder" {>= "1.0+beta10"}
   "ipaddr" {>= "2.6.0" & < "5.0.0"}
-  "functoria" {>= "2.2.0"}
+  "functoria" {>= "2.2.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.2.0/opam:23]:
```suggestion
  "functoria" {>= "2.2.0" & < "4.0"}
```

File: packages/mirage/mirage.3.3.0/opam [hunk 10]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.04.2"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "2.6.0" & < "5.0.0"}
-  "functoria"          {>= "2.2.2"}
+  "functoria"          {>= "2.2.2" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.3.0/opam:23]:
```suggestion
  "functoria"          {>= "2.2.2" & < "4.0"}
```

File: packages/mirage/mirage.3.3.1/opam [hunk 11]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.04.2"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "2.6.0" & < "5.0.0"}
-  "functoria"          {>= "2.2.3"}
+  "functoria"          {>= "2.2.3" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.3.1/opam:23]:
```suggestion
  "functoria"          {>= "2.2.3" & < "4.0"}
```

File: packages/mirage/mirage.3.4.1/opam [hunk 12]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.04.2"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "2.2.3"}
+  "functoria"          {>= "2.2.3" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.4.1/opam:23]:
```suggestion
  "functoria"          {>= "2.2.3" & < "4.0"}
```

File: packages/mirage/mirage.3.5.1/opam [hunk 13]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.05.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "2.2.3"}
+  "functoria"          {>= "2.2.3" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.5.1/opam:23]:
```suggestion
  "functoria"          {>= "2.2.3" & < "4.0"}
```

File: packages/mirage/mirage.3.5.2/opam [hunk 14]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.05.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "2.2.3"}
+  "functoria"          {>= "2.2.3" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.5.2/opam:23]:
```suggestion
  "functoria"          {>= "2.2.3" & < "4.0"}
```

File: packages/mirage/mirage.3.6.0/opam [hunk 15]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.05.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "2.2.3"}
+  "functoria"          {>= "2.2.3" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.6.0/opam:23]:
```suggestion
  "functoria"          {>= "2.2.3" & < "4.0"}
```

File: packages/mirage/mirage.3.7.1/opam [hunk 16]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.06.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "3.0.2"}
+  "functoria"          {>= "3.0.2" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.7.1/opam:23]:
```suggestion
  "functoria"          {>= "3.0.2" & < "4.0"}
```

File: packages/mirage/mirage.3.7.2/opam [hunk 17]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.06.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "3.0.2"}
+  "functoria"          {>= "3.0.2" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.7.2/opam:23]:
```suggestion
  "functoria"          {>= "3.0.2" & < "4.0"}
```

File: packages/mirage/mirage.3.7.3/opam [hunk 18]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.06.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "3.0.2"}
+  "functoria"          {>= "3.0.2" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.7.3/opam:23]:
```suggestion
  "functoria"          {>= "3.0.2" & < "4.0"}
```

File: packages/mirage/mirage.3.7.4/opam [hunk 19]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.06.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "3.0.2"}
+  "functoria"          {>= "3.0.2" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.7.4/opam:23]:
```suggestion
  "functoria"          {>= "3.0.2" & < "4.0"}
```

File: packages/mirage/mirage.3.7.6/opam [hunk 20]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.06.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.7.6/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.7.7/opam [hunk 21]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.06.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "3.0.0" & < "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.7.7/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.8.1/opam [hunk 22]
```diff
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.06.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.8.1/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

File: packages/mirage/mirage.3.9.0/opam [hunk 23]
```diff
@@ -20,8 +20,9 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "1.1.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0"}
+  "functoria"          {>= "3.1.0" & < "4.0.0"}
```
Comment (hannesm) [inline packages/mirage/mirage.3.9.0/opam:23]:
```suggestion
  "functoria"          {>= "3.1.0" & < "4.0"}
```

**Comment by hannesm** [comment]
I took the liberty to push another commit that adds a few more upper bounds to the mirage packages.

**Comment by hannesm** [comment]
looks mostly good, though conan-cli fails (and I've no clue why) -- also given the size of the CI jobs, maybe worth to defer the conan-cli fix to a different PR?

```
# (cd _build/.sandbox/124e9214e57385d40a71223c3d1b0669/default && /usr/bin/git --no-pager diff --no-index --color=always -u ../../../default/test/unikernel.t test/unikernel.t.corrected)
# diff --git a/../../../default/test/unikernel.t b/test/unikernel.t.corrected
# index b846f8c..c47de06 100644
# --- a/../../../default/test/unikernel.t
# +++ b/test/unikernel.t.corrected
# @@ -2,5 +2,7 @@ Simple unikernel with conan
#    $ cd ../unikernel
#    $ mirage configure -t unix
#    $ mirage build 2> /dev/null > /dev/null
# +  [1]
#    $ ./main
# -  MIME type: application/zlib
# +  ./main: not found
# +  [127]
```

**Comment by hannesm** [comment]
I looked through the failures, and apart from the conan-cli mentioned above, mirage 3.10.5 needs a patch (but I hesitate to push to this PR and trigger another 10000 builds):
```patch
diff --git a/packages/mirage/mirage.3.10.5/opam b/packages/mirage/mirage.3.10.5/opam
index 580a0d9ef4..497dd385af 100644
--- a/packages/mirage/mirage.3.10.5/opam
+++ b/packages/mirage/mirage.3.10.5/opam
@@ -20,7 +20,7 @@ depends: [
   "ocaml" {>= "4.08.0"}
   "dune" {>= "2.0"}
   "ipaddr"             {>= "5.0.0"}
-  "functoria"          {>= "3.1.0" & < "4.0.0"}
+  "functoria"          {>= "3.1.0" & < "4.0"}
   "bos"
   "astring"
   "logs"
```

**Comment by dinosaure** [comment]
I added your patch because we know that we must do that but yeah, the `opam-ci` becomes insanely huge. @kit-ty-kate do you have any opinions on this PR?

**Comment by dinosaure** [comment]
I think this PR seems good enough :+1:.

**Comment by samoht** [comment]
Could you move the mirage3 bounds to another PR so that we can merge this directly as this is a direct improvement to the quality of this repository?

Regarding the current PR: I have some doubts about using  `available: opam-version >= "2.1.0"` in the opam files of mirage4 as this breaks `ocaml-ci`. Can't we just use `opam-version: "2.1"` + `flags: avoid-version` as you suggested?

**Comment by samoht** [comment]
(I'll open the new PR for the Mirage3 bounds)

**Comment by samoht** [comment]
I'm temporary closing this PR as opam-repo-ci seems completely stuck

**Comment by hannesm** [comment]
Thanks @samoht. I also extracted some PRs, so with
- #20581 (functoria-runtime)
- #20582 (mirage-crypto-pk)
- #20583 (conan)
- #20584 (mirage 3)

the only change needed is the mirage-runtime packages (upper bounds on functoria-runtime) as ar as I can tell.

**Comment by samoht** [comment]
I'm re-opening that one now - hopefully the CI will pick it up (if not I'll open a new one)

**Comment by samoht** [comment]
Also please wait before merging this PR, there are a few more things I'd like to check first and I'd like to change the tag upstream as well.

**Comment by samoht** [comment]
Superseeded by #20588 (I've changed the tarball)

**Pull request closed without merging**